File: fe2-android/app/src/test/unit/java/com/github/dedis/popstellar/utility/security/KeyManagerTest.java
package com.github.dedis.popstellar.utility.security;

import androidx.arch.core.executor.testing.InstantTaskExecutorRule;
import androidx.test.ext.junit.runners.AndroidJUnit4;

import com.github.dedis.popstellar.model.objects.*;
import com.github.dedis.popstellar.model.objects.event.EventState;
import com.github.dedis.popstellar.model.objects.security.*;
import com.github.dedis.popstellar.repository.RollCallRepository;
import com.github.dedis.popstellar.testutils.Base64DataUtils;
import com.github.dedis.popstellar.utility.error.keys.*;
import com.google.crypto.tink.PublicKeySign;
import com.google.crypto.tink.integration.android.AndroidKeysetManager;

import org.junit.*;
import org.junit.rules.RuleChain;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnit;

import java.security.GeneralSecurityException;
import java.util.HashSet;

import javax.inject.Inject;

import dagger.hilt.android.testing.HiltAndroidRule;
import dagger.hilt.android.testing.HiltAndroidTest;

import static com.github.dedis.popstellar.di.KeysetModule.DeviceKeyset;
import static org.junit.Assert.*;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.*;

@HiltAndroidTest
@RunWith(AndroidJUnit4.class)
public class KeyManagerTest {

  private final HiltAndroidRule hiltRule = new HiltAndroidRule(this);

  @Rule public RuleChain rule = RuleChain.outerRule(hiltRule).around(MockitoJUnit.testRule(this));

  @Rule public InstantTaskExecutorRule executorRule = new InstantTaskExecutorRule();

  @Inject @DeviceKeyset AndroidKeysetManager androidKeysetManager;
  @Inject RollCallRepository rollCallRepo;
  @Mock private Wallet wallet;

  @Before
  public void setup() {
    hiltRule.inject();
  }

  @Test
  public void keyPairIsRight() throws GeneralSecurityException {
    KeyManager keyManager = new KeyManager(androidKeysetManager, wallet);
    KeyPair mainKeyPair = keyManager.getMainKeyPair();
    PublicKey mainKey = keyManager.getMainPublicKey();

    assertEquals(mainKey, mainKeyPair.getPublicKey());

    // We cannot extract the public key from the keyset handle
    // But we can make sure both signatures are equals and the key manager keypair can verify it
    PublicKeySign signer = androidKeysetManager.getKeysetHandle().getPrimitive(PublicKeySign.class);
    // Generate any data, here a message id
    Base64URLData data = Base64DataUtils.generateMessageID();

    Signature signature = mainKeyPair.sign(data);
    assertArrayEquals(signer.sign(data.getData()), signature.getData());
    assertTrue(mainKey.verify(signature, data));
  }

  @Test
  public void popTokenRetrievingWorks() throws KeyException {
    PoPToken token = Base64DataUtils.generatePoPToken();
    when(wallet.recoverKey(any(), any(), any())).thenReturn(token);

    // create LAO and RollCalls
    Lao lao = new Lao("lao", Base64DataUtils.generatePublicKey(), 54213424);
    String rollCallName1 = "rollcall1";
    String rollCallName2 = "rollcall2";
    long creation1 = 5421364;
    long creation2 = 5421363;
    String id1 = RollCall.generateCreateRollCallId(lao.getId(), creation1, rollCallName1);
    String id2 = RollCall.generateCreateRollCallId(lao.getId(), creation2, rollCallName2);
    RollCall rollCall1 =
        new RollCall(
            id1,
            id1,
            rollCallName1,
            creation1,
            creation1 + 1,
            creation1 + 75,
            EventState.CLOSED,
            new HashSet<>(),
            "location",
            "desc");
    RollCall rollCall2 =
        new RollCall(
            id2,
            id2,
            rollCallName2,
            creation2,
            creation2 + 1,
            creation2 + 75,
            EventState.CLOSED,
            new HashSet<>(),
            "EPFL",
            "do not come");

    rollCallRepo.updateRollCall(lao.getId(), rollCall1, rollCall1.getId());
    rollCallRepo.updateRollCall(lao.getId(), rollCall2, rollCall2.getId());

    KeyManager manager = new KeyManager(androidKeysetManager, wallet);
    assertEquals(
        token,
        manager.getValidPoPToken(lao.getId(), rollCallRepo.getLastClosedRollCall(lao.getId())));
    assertEquals(token, manager.getValidPoPToken(lao.getId(), rollCall1));

    // make sure that rollcall1 was taken and not rollcall2 as the oldest is rollcall 1
    verify(wallet, atLeast(1)).recoverKey(eq(lao.getId()), eq(rollCall1.getId()), any());
  }

  @Test
  public void popTokenRetrievingFailsWhenWalletFails() throws KeyException {
    PoPToken token = Base64DataUtils.generatePoPToken();

    // create LAO and RollCall
    Lao lao = new Lao("lao", Base64DataUtils.generatePublicKey(), 54213424);
    String id = RollCall.generateCreateRollCallId(lao.getId(), 5421364, "rollcall");
    RollCall rollCall =
        new RollCall(
            id,
            id,
            "rollcall",
            5421364,
            5421364 + 1,
            5421364 + 145,
            EventState.CLOSED,
            new HashSet<>(),
            "ETHZ",
            "do come");

    rollCallRepo.updateRollCall(lao.getId(), rollCall, rollCall.getId());
    KeyManager manager = new KeyManager(androidKeysetManager, wallet);

    // Test with every possible errors
    when(wallet.recoverKey(any(), any(), any()))
        .thenThrow(new KeyGenerationException(new GeneralSecurityException()));
    assertThrows(
        KeyGenerationException.class,
        () ->
            manager.getValidPoPToken(lao.getId(), rollCallRepo.getLastClosedRollCall(lao.getId())));
    verify(wallet, times(1)).recoverKey(eq(lao.getId()), eq(rollCall.getId()), any());
    reset(wallet);

    when(wallet.recoverKey(any(), any(), any())).thenThrow(new UninitializedWalletException());
    assertThrows(
        UninitializedWalletException.class,
        () ->
            manager.getValidPoPToken(lao.getId(), rollCallRepo.getLastClosedRollCall(lao.getId())));
    verify(wallet, times(1)).recoverKey(eq(lao.getId()), eq(rollCall.getId()), any());
    reset(wallet);

    when(wallet.recoverKey(any(), any(), any())).thenThrow(new InvalidPoPTokenException(token));
    assertThrows(
        InvalidPoPTokenException.class,
        () ->
            manager.getValidPoPToken(lao.getId(), rollCallRepo.getLastClosedRollCall(lao.getId())));
    verify(wallet, times(1)).recoverKey(eq(lao.getId()), eq(rollCall.getId()), any());
  }

  @Test
  public void popTokenRetrievingFailsWhenLaoHasNoRollCall() {
    // create LAO
    Lao lao = new Lao("lao", Base64DataUtils.generatePublicKey(), 54213424);

    KeyManager manager = new KeyManager(androidKeysetManager, wallet);
    assertThrows(
        NoRollCallException.class,
        () ->
            manager.getValidPoPToken(lao.getId(), rollCallRepo.getLastClosedRollCall(lao.getId())));
  }
}

File: fe2-android/app/src/test/unit/java/com/github/dedis/popstellar/utility/security/KeyManagerTest.kt
package com.github.dedis.popstellar.utility.security

import androidx.arch.core.executor.testing.InstantTaskExecutorRule
import androidx.test.ext.junit.runners.AndroidJUnit4
import com.github.dedis.popstellar.di.KeysetModule.DeviceKeyset
import com.github.dedis.popstellar.model.objects.Lao
import com.github.dedis.popstellar.model.objects.RollCall
import com.github.dedis.popstellar.model.objects.RollCall.Companion.generateCreateRollCallId
import com.github.dedis.popstellar.model.objects.Wallet
import com.github.dedis.popstellar.model.objects.event.EventState
import com.github.dedis.popstellar.model.objects.security.Base64URLData
import com.github.dedis.popstellar.repository.RollCallRepository
import com.github.dedis.popstellar.testutils.Base64DataUtils
import com.github.dedis.popstellar.testutils.MockitoKotlinHelpers
import com.github.dedis.popstellar.utility.error.keys.InvalidPoPTokenException
import com.github.dedis.popstellar.utility.error.keys.KeyException
import com.github.dedis.popstellar.utility.error.keys.KeyGenerationException
import com.github.dedis.popstellar.utility.error.keys.NoRollCallException
import com.github.dedis.popstellar.utility.error.keys.UninitializedWalletException
import com.google.crypto.tink.PublicKeySign
import com.google.crypto.tink.integration.android.AndroidKeysetManager
import dagger.hilt.android.testing.HiltAndroidRule
import dagger.hilt.android.testing.HiltAndroidTest
import java.security.GeneralSecurityException
import javax.inject.Inject
import org.junit.Assert
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.rules.RuleChain
import org.junit.runner.RunWith
import org.mockito.Mock
import org.mockito.Mockito
import org.mockito.junit.MockitoJUnit

@HiltAndroidTest
@RunWith(AndroidJUnit4::class)
class KeyManagerTest {
  private val hiltRule = HiltAndroidRule(this)

  @JvmField
  @Rule
  var rule: RuleChain = RuleChain.outerRule(hiltRule).around(MockitoJUnit.testRule(this))

  @JvmField @Rule var executorRule = InstantTaskExecutorRule()

  @Inject @DeviceKeyset lateinit var androidKeysetManager: AndroidKeysetManager

  @Inject lateinit var rollCallRepo: RollCallRepository

  @Mock private lateinit var wallet: Wallet

  @Before
  fun setup() {
    hiltRule.inject()
  }

  @Test
  @Throws(GeneralSecurityException::class)
  fun keyPairIsRight() {
    val keyManager = KeyManager(androidKeysetManager, wallet)
    val mainKeyPair = keyManager.mainKeyPair
    val mainKey = keyManager.mainPublicKey

    Assert.assertEquals(mainKey, mainKeyPair.publicKey)

    // We cannot extract the public key from the keyset handle
    // But we can make sure both signatures are equals and the key manager keypair can verify it
    val signer = androidKeysetManager.keysetHandle.getPrimitive(PublicKeySign::class.java)

    // Generate any data, here a message id
    val data: Base64URLData = Base64DataUtils.generateMessageID()
    val signature = mainKeyPair.sign(data)

    Assert.assertArrayEquals(signer.sign(data.data), signature.data)
    Assert.assertTrue(mainKey.verify(signature, data))
  }

  @Test
  @Throws(KeyException::class)
  fun popTokenRetrievingWorks() {
    val token = Base64DataUtils.generatePoPToken()
    Mockito.`when`(
        wallet.recoverKey(
          MockitoKotlinHelpers.any(),
          MockitoKotlinHelpers.any(),
          MockitoKotlinHelpers.any()
        )
      )
      .thenReturn(token)

    // create LAO and RollCalls
    val lao = Lao("lao", Base64DataUtils.generatePublicKey(), 54213424)
    val rollCallName1 = "rollcall1"
    val rollCallName2 = "rollcall2"
    val creation1: Long = 5421364
    val creation2: Long = 5421363
    val id1 = generateCreateRollCallId(lao.id, creation1, rollCallName1)
    val id2 = generateCreateRollCallId(lao.id, creation2, rollCallName2)
    val rollCall1 =
      RollCall(
        id1,
        id1,
        rollCallName1,
        creation1,
        creation1 + 1,
        creation1 + 75,
        EventState.CLOSED,
        LinkedHashSet(),
        "location",
        "desc"
      )
    val rollCall2 =
      RollCall(
        id2,
        id2,
        rollCallName2,
        creation2,
        creation2 + 1,
        creation2 + 75,
        EventState.CLOSED,
        LinkedHashSet(),
        "EPFL",
        "do not come"
      )
    rollCallRepo.updateRollCall(lao.id, rollCall1)
    rollCallRepo.updateRollCall(lao.id, rollCall2)

    val manager = KeyManager(androidKeysetManager, wallet)

    Assert.assertEquals(
      token,
      manager.getValidPoPToken(lao.id, rollCallRepo.getLastClosedRollCall(lao.id))
    )
    Assert.assertEquals(token, manager.getValidPoPToken(lao.id, rollCall1))

    // make sure that rollcall1 was taken and not rollcall2 as the oldest is rollcall 1
    Mockito.verify(wallet, Mockito.atLeast(1))
      .recoverKey(
        MockitoKotlinHelpers.eq(lao.id),
        MockitoKotlinHelpers.eq(rollCall1.id),
        MockitoKotlinHelpers.any()
      )
  }

  @Test
  @Throws(KeyException::class)
  fun popTokenRetrievingFailsWhenWalletFails() {
    val token = Base64DataUtils.generatePoPToken()

    // create LAO and RollCall
    val lao = Lao("lao", Base64DataUtils.generatePublicKey(), 54213424)
    val id = generateCreateRollCallId(lao.id, 5421364, "rollcall")
    val rollCall =
      RollCall(
        id,
        id,
        "rollcall",
        5421364,
        (5421364 + 1).toLong(),
        (5421364 + 145).toLong(),
        EventState.CLOSED,
        LinkedHashSet(),
        "ETHZ",
        "do come"
      )
    rollCallRepo.updateRollCall(lao.id, rollCall)

    val manager = KeyManager(androidKeysetManager, wallet)
    // Test with every possible errors
    Mockito.`when`(
        wallet.recoverKey(
          MockitoKotlinHelpers.any(),
          MockitoKotlinHelpers.any(),
          MockitoKotlinHelpers.any()
        )
      )
      .thenThrow(KeyGenerationException(GeneralSecurityException()))

    Assert.assertThrows(KeyGenerationException::class.java) {
      manager.getValidPoPToken(lao.id, rollCallRepo.getLastClosedRollCall(lao.id))
    }
    Mockito.verify(wallet, Mockito.times(1))
      .recoverKey(
        MockitoKotlinHelpers.eq(lao.id),
        MockitoKotlinHelpers.eq(rollCall.id),
        MockitoKotlinHelpers.any()
      )

    Mockito.reset(wallet)

    Mockito.`when`(
        wallet.recoverKey(
          MockitoKotlinHelpers.any(),
          MockitoKotlinHelpers.any(),
          MockitoKotlinHelpers.any()
        )
      )
      .thenThrow(UninitializedWalletException())

    Assert.assertThrows(UninitializedWalletException::class.java) {
      manager.getValidPoPToken(lao.id, rollCallRepo.getLastClosedRollCall(lao.id))
    }
    Mockito.verify(wallet, Mockito.times(1))
      .recoverKey(
        MockitoKotlinHelpers.eq(lao.id),
        MockitoKotlinHelpers.eq(rollCall.id),
        MockitoKotlinHelpers.any()
      )

    Mockito.reset(wallet)

    Mockito.`when`(
        wallet.recoverKey(
          MockitoKotlinHelpers.any(),
          MockitoKotlinHelpers.any(),
          MockitoKotlinHelpers.any()
        )
      )
      .thenThrow(InvalidPoPTokenException(token))

    Assert.assertThrows(InvalidPoPTokenException::class.java) {
      manager.getValidPoPToken(lao.id, rollCallRepo.getLastClosedRollCall(lao.id))
    }
    Mockito.verify(wallet, Mockito.times(1))
      .recoverKey(
        MockitoKotlinHelpers.eq(lao.id),
        MockitoKotlinHelpers.eq(rollCall.id),
        MockitoKotlinHelpers.any()
      )
  }

  @Test
  fun popTokenRetrievingFailsWhenLaoHasNoRollCall() {
    // create LAO
    val lao = Lao("lao", Base64DataUtils.generatePublicKey(), 54213424)
    val manager = KeyManager(androidKeysetManager, wallet)
    Assert.assertThrows(NoRollCallException::class.java) {
      manager.getValidPoPToken(lao.id, rollCallRepo.getLastClosedRollCall(lao.id))
    }
  }
}
